### PR TITLE
Added main ToS/PP

### PR DESCRIPTION
**Pull request by lsirac**
Changes toward GDPR compliance (#1245)

Still need to add a ToS/Pp footer nearly everywhere.

### CODE REVIEW

**Comment by samtstern** [comment]
@SUPERCILEX FYI: @lsirac is a new member of the Firebase Auth team and is going to be helping out on FirebaseUI, changing the ToS/PP flow to be compliant is his first task.

File: auth/src/main/java/com/firebase/ui/auth/util/ui/PreambleHandler.java [hunk 1]
```diff
@@ -55,8 +74,9 @@ private void setPreamble(TextView textView) {
         textView.setText(mBuilder);
     }
 
-    private void setupCreateAccountPreamble() {
-        String withTargets = getPreambleStringWithTargets();
+    private void setupPreamble(@StringRes int textViewText) {
```
Comment (samtstern) [inline auth/src/main/java/com/firebase/ui/auth/util/ui/PreambleHandler.java]: nit: don't really like this method name because `setupPreamble` and `setPreamble` are so similar.

File: auth/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -5,6 +5,8 @@
     <string name="fui_default_toolbar_title" translatable="false">@string/app_name</string>
     <string name="fui_progress_dialog_loading" translation_description="Loading text in dialog">Loading…</string>
     <string name="fui_sign_in_default" translation_description="Button text to sign in">Sign in</string>
+    <string name="fui_tos_and_pp" translation_description="Global ToS and Privacy policy message">By continuing, you are indicating that you accept our <xliff:g example="https://google.com/terms" id="tos" translation_description="">%1$s</xliff:g> and <xliff:g example="https://google.com/privacy" id="pp" translation_description="">%2$s</xliff:g>.</string>
```
Comment (samtstern) [inline auth/src/main/res/values/strings.xml]: If you think this text is final let me know and I will ship them off for translation.

**Comment by samtstern** [comment]
Some images, for easy commenting:

**NASCAR**
</img>

**Check Email**
</img>

**Phone Auth, Long**
</img>

**Phone Auth, Short**
</img>

File: auth/src/main/java/com/firebase/ui/auth/data/model/FlowParameters.java [hunk 3]
```diff
@@ -128,4 +131,24 @@ public FlowParameters createFromParcel(Parcel in) {
...
+
+    public @StringRes int getGlobalTermsStringResource() {
+        boolean hasTos = !TextUtils.isEmpty(termsOfServiceUrl);
+        boolean hasPp = !TextUtils.isEmpty(privacyPolicyUrl);
+        if(hasTos && hasPp) {
+            return R.string.fui_tos_and_pp;
+        } else if(hasTos) {
+            return R.string.fui_tos_only;
+        } else {
+            return R.string.fui_pp_only;
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/data/model/FlowParameters.java]: This else implies the default is the privacy policy—is that what you want? If so, we should add checks, update the README n' docs, throw exceptions, yada yada for `AuthUI` to ensure the developer must supply either a tos or a pp. Otherwise, we should return -1 and handle that case.

File: auth/src/main/java/com/firebase/ui/auth/data/model/FlowParameters.java [hunk 4]
```diff
@@ -128,4 +131,24 @@ public FlowParameters createFromParcel(Parcel in) {
...
+        }
+    }
+
+    public @StringRes int getGlobalTermsFooterStringResource() {
+        return R.string.fui_tos_and_pp_footer;
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/data/model/FlowParameters.java]: Should this have if statements like above? If not, let's just inline it in the XML.

File: auth/src/main/java/com/firebase/ui/auth/ui/email/EmailActivity.java [hunk 5]
```diff
@@ -130,4 +134,14 @@ private void setSlideAnimation() {
         // Make the next activity slide in
         overridePendingTransition(R.anim.fui_slide_in_right, R.anim.fui_slide_out_left);
     }
+
+    private void setUpTermsOfServiceFooter() {
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/ui/email/EmailActivity.java]: I'm seeing this method a lot. I wonder if it makes sense to add a `PrivacyDisclosureUtils` class that includes this and the tos methods you added in `FlowParameters` (I'm not a huge fan of logic in a POJO). @samtstern What do you think?

File: auth/src/main/java/com/firebase/ui/auth/ui/phone/PhoneActivity.java [hunk 6]
```diff
@@ -406,4 +409,14 @@ private SubmitConfirmationCodeFragment getSubmitConfirmationCodeFragment() {
         return (SubmitConfirmationCodeFragment) getSupportFragmentManager().findFragmentByTag
                 (SubmitConfirmationCodeFragment.TAG);
     }
+
+    private void setUpTermsOfServiceFooter() {
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/ui/phone/PhoneActivity.java]: Yeah, I'm definitely thinking we need a utility, this is numeros quatros I've seen. 😆

File: auth/src/main/java/com/firebase/ui/auth/util/ui/PreambleHandler.java [hunk 7]
```diff
@@ -25,6 +25,7 @@
     private static final String BTN_TARGET = "%BTN%";
     private static final String TOS_TARGET = "%TOS%";
     private static final String PP_TARGET = "%PP%";
+    private static final int NO_BUTTON = -1;
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/util/ui/PreambleHandler.java]: See my patch for improvements to this class.

File: auth/src/main/res/layout-land/fui_auth_method_picker_layout.xml [hunk 8]
```diff
@@ -32,6 +32,16 @@
             android:id="@+id/btn_holder"
             style="@style/FirebaseUI.AuthMethodPicker.ButtonHolder" />
 
+        <TextView
```
Comment (SUPERCILEX) [inline auth/src/main/res/layout-land/fui_auth_method_picker_layout.xml]: I believe you meant to put this out of the scroll view, right? If not, you can't put more than one view in a `ScrollView` so we'll need a LL.

File: auth/src/main/res/layout/fui_activity_register_email.xml [hunk 9]
```diff
@@ -5,4 +5,14 @@
...
     android:layout_height="match_parent"
-    tools:ignore="MergeRootFrame" />
+    tools:ignore="MergeRootFrame" >
+
+    <TextView
+        android:id="@+id/email_footer_tos_and_pp_text"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="@dimen/fui_field_padding_vert"
+        android:layout_gravity="bottom|right"
```
Comment (SUPERCILEX) [inline auth/src/main/res/layout/fui_activity_register_email.xml]: Let's use `end` instead of `right`.

File: auth/src/main/res/layout/fui_auth_method_picker_layout.xml [hunk 10]
```diff
@@ -26,12 +27,23 @@
...
+            style="@style/FirebaseUI.AuthMethodPicker.ButtonHolder"/>
     </ScrollView>
-
+    <TextView
+        android:id="@+id/main_tos_and_pp"
+        style="@style/FirebaseUI.Text.BodyText"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="@dimen/fui_field_padding_vert"
+        android:textColor="?android:textColorTertiary"
```
Comment (SUPERCILEX) [inline auth/src/main/res/layout/fui_auth_method_picker_layout.xml]: Does the color change if you remove this? Either way, we need to use some AppCompat version, not the system one. I'll get back to you if the color is wrong when you remove it.

File: auth/src/main/res/values/strings.xml [hunk 11]
```diff
@@ -100,4 +105,8 @@
     <string name="fui_verify_phone_number" translation_description="Button text to submit confirmation code and complete phone verification">Verify Phone Number</string>
     <string name="fui_continue_phone_login" translation_description="Button text to submit phone number and send sms">Continue</string>
     <string name="fui_sms_terms_of_service" translation_description="Fine print warning displayed below Verify Phone Number button">By tapping “%1$s”, an SMS may be sent. Message &amp; data rates may apply.</string>
+    <string name="fui_sms_terms_of_service_and_privacy_policy_extended" translation_description="Fine print warning displayed below Verify Phone Number button">By tapping “%1$s”, you are indicating that you accept our <xliff:g example="https://google.com/terms" id="tos" translation_description="">%2$s</xliff:g> and <xliff:g example="https://google.com/privacy" id="pp" translation_description="">%3$s</xliff:g>. An SMS may be sent. Message &amp; data rates may apply.</string>
```
Comment (SUPERCILEX) [inline auth/src/main/res/values/strings.xml]: We're going to need to kill `fui_create_account_preamble_tos_and_pp` and the others to make lint happy. In the meantime, add `'MissingTranslation' // TODO remove in future PR` [here](https://github.com/SUPERCILEX/FirebaseUI-Android/blob/4e51a457b12e622604e61b473079b2107a54cfd7/library/quality/quality.gradle#L76-L79) and then @samtstern will probably have you create a second PR that kills those strings.

File: auth/src/main/java/com/firebase/ui/auth/ui/email/EmailActivity.java (hunk 5 above)
Comment (samtstern) [reply auth/src/main/java/com/firebase/ui/auth/ui/email/EmailActivity.java]: Seems like a good idea to me, less repetition is better!

File: auth/src/main/res/values/strings.xml [hunk 12]
```diff
@@ -5,6 +5,11 @@
     <string name="fui_default_toolbar_title" translatable="false">@string/app_name</string>
     <string name="fui_progress_dialog_loading" translation_description="Loading text in dialog">Loading…</string>
     <string name="fui_sign_in_default" translation_description="Button text to sign in">Sign in</string>
+    <string name="fui_tos_and_pp" translation_description="Global ToS and Privacy policy message">By continuing, you are indicating that you accept our <xliff:g example="https://google.com/terms" id="tos" translation_description="">%1$s</xliff:g> and <xliff:g example="https://google.com/privacy" id="pp" translation_description="">%2$s</xliff:g>.</string>
```
Comment (samtstern) [inline auth/src/main/res/values/strings.xml]: Also make sure you add some translation_description that's not ""

File: auth/src/main/java/com/firebase/ui/auth/data/model/FlowParameters.java (hunk 3 above)
Comment (lsirac) [reply auth/src/main/java/com/firebase/ui/auth/data/model/FlowParameters.java]: You're right, I'll change this. We won't force them to have a ToS/Privacy policy though.

File: auth/src/main/java/com/firebase/ui/auth/data/model/FlowParameters.java (hunk 4 above)
Comment (lsirac) [reply auth/src/main/java/com/firebase/ui/auth/data/model/FlowParameters.java]: I added 2 more strings to deal with ToS only & Pp only. Having this makes more sense now :)

File: auth/src/main/java/com/firebase/ui/auth/ui/phone/PhoneActivity.java (hunk 6 above)
Comment (lsirac) [reply auth/src/main/java/com/firebase/ui/auth/ui/phone/PhoneActivity.java]: :sweat_smile:

File: auth/src/main/java/com/firebase/ui/auth/data/model/FlowParameters.java [hunk 13]
```diff
@@ -128,4 +131,24 @@ public FlowParameters createFromParcel(Parcel in) {
             return new FlowParameters[size];
         }
     };
+
+    public boolean isSingleProviderFlow() {
```
[the remark this answers is not in the document]
Comment (lsirac) [reply auth/src/main/java/com/firebase/ui/auth/data/model/FlowParameters.java]: Sure

File: auth/src/main/java/com/firebase/ui/auth/ui/email/RegisterEmailFragment.java [hunk 14]
```diff
@@ -78,7 +80,9 @@ public void onCreate(@Nullable Bundle savedInstanceState) {
             @Override
             protected void onSuccess(@NonNull IdpResponse response) {
                 startSaveCredentials(
-                        mHandler.getCurrentUser(), response, mPasswordEditText.getText().toString());
+                        mHandler.getCurrentUser(),
```
Comment (lsirac) [inline auth/src/main/java/com/firebase/ui/auth/ui/email/RegisterEmailFragment.java]: Is something wrong with the formatting settings?

File: auth/src/main/java/com/firebase/ui/auth/ui/email/RegisterEmailFragment.java [hunk 15]
```diff
@@ -226,11 +233,11 @@ private void validateAndRegisterUser() {
         boolean nameValid = mNameValidator.validate(name);
         if (emailValid && passwordValid && nameValid) {
             mHandler.startSignIn(new IdpResponse.Builder(
-                    new User.Builder(EmailAuthProvider.PROVIDER_ID, email)
-                            .setName(name)
-                            .setPhotoUri(mUser.getPhotoUri())
-                            .build())
-                    .build(),
+                            new User.Builder(EmailAuthProvider.PROVIDER_ID, email)
```
Comment (lsirac) [inline auth/src/main/java/com/firebase/ui/auth/ui/email/RegisterEmailFragment.java]: Is something wrong with the formatting settings?

Comment (samtstern) [reply auth/src/main/java/com/firebase/ui/auth/ui/email/RegisterEmailFragment.java]: I wouldn't worry about this.  Equally legible either way.

Comment (SUPERCILEX) [reply auth/src/main/java/com/firebase/ui/auth/ui/email/RegisterEmailFragment.java]: Nope, we just made an indentation change to the styling a long time ago and never bothered running it over the whole codebase. Like @samtstern said, we're not _that_ nit picky. 😆 Thanks for checking though! 👍

I'll take another look at this tomorrow.

File: library/quality/quality.gradle [hunk 16]
```diff
@@ -77,6 +77,7 @@ android {
                 'IconExpectedSize',
                 'InvalidPackage', // Firestore uses GRPC which makes lint mad
                 'NewerVersionAvailable', 'GradleDependency' // For reproducible builds
+                'MissingTranslation' // TODO remove in future PR
```
Comment (SUPERCILEX) [inline library/quality/quality.gradle]: I think you're also going to have to add `UnusedResources`. And you need a comma above. 😄

File: auth/src/main/java/com/firebase/ui/auth/ui/email/CheckEmailFragment.java [hunk 17]
```diff
@@ -96,6 +99,19 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
         }
 
         view.findViewById(R.id.button_next).setOnClickListener(this);
+
+        TextView termsText = view.<TextView>findViewById(R.id.email_tos_and_pp_text);
```
Comment (samtstern) [inline auth/src/main/java/com/firebase/ui/auth/ui/email/CheckEmailFragment.java]: optional style nit:

I normally see this:
```java
TextView termsText = (TextView) view.findViewById(R.id.email_tos_and_pp_text);
```

Not this:
```java
TextView termsText = view.<TextView>findViewById(R.id.email_tos_and_pp_text);
```

File: auth/src/main/java/com/firebase/ui/auth/ui/idp/AuthMethodPickerActivity.java [hunk 18]
```diff
@@ -106,6 +108,11 @@ protected void onFailure(@NonNull Exception e) {
                 }
             }
         });
+
+        TextView termsText = findViewById(R.id.main_tos_and_pp);
+        PrivacyDisclosureUtils.setupTermsOfServiceAndPrivacyPolicyText(AuthMethodPickerActivity.this,
```
Comment (samtstern) [inline auth/src/main/java/com/firebase/ui/auth/ui/idp/AuthMethodPickerActivity.java]: nit: do you need `AuthMethodPickerActivity.this` or could it be just `this`?

File: auth/src/main/java/com/firebase/ui/auth/ui/phone/SubmitConfirmationCodeFragment.java [hunk 19]
```diff
@@ -73,8 +72,10 @@ public static SubmitConfirmationCodeFragment newInstance(FlowParameters flowPara
 
     @Nullable
     @Override
-    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable
-            Bundle savedInstanceState) {
+    public View onCreateView(@NonNull LayoutInflater inflater,
+                             @Nullable ViewGroup container,
+                             @Nullable
```
Comment (samtstern) [inline auth/src/main/java/com/firebase/ui/auth/ui/phone/SubmitConfirmationCodeFragment.java]: nit: formatter kinda nuked this one.  Can we put the `Bundle savedInstanceState` back on the line with `@Nullable`?

File: auth/src/main/java/com/firebase/ui/auth/ui/phone/VerifyPhoneNumberFragment.java [hunk 20]
```diff
@@ -273,6 +272,23 @@ private void setCountryCode(PhoneNumber phoneNumber) {
...
+        final String verifyPhoneButtonText = getString(R.string.fui_verify_phone_number);
+        final String multipleProviderFlowText = getString(R.string.fui_sms_terms_of_service,
+                verifyPhoneButtonText);
+        FlowParameters flowParameters = getFlowParams();
+        PrivacyDisclosureUtils.setupTermsOfServiceAndPrivacyPolicySmsText(getContext(),
+                flowParameters,
+                mSmsTermsText,
+                multipleProviderFlowText);
+
+        if (!flowParameters.isSingleProviderFlow()) {
```
Comment (samtstern) [inline auth/src/main/java/com/firebase/ui/auth/ui/phone/VerifyPhoneNumberFragment.java]: So if this **is** a single provider flow then we'll have both the big text and the footer?  If that's not the case (maybe the previous method has some internal switching) let's make that clearer here by doing `if / else` so only one of the `PrivacyDisclosureUtils` classes runs.

File: auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+    public static void setupTermsOfServiceFooter(Context context,
+                                                 FlowParameters flowParameters,
+                                                 TextView footerText) {
+        PreambleHandler.setup(context,
+                flowParameters,
+                getGlobalTermsFooterStringResource(flowParameters),
+                footerText);
+    }
+
+    public static void setupTermsOfServiceAndPrivacyPolicySmsText(Context context,
```
Comment (samtstern) [inline auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java]: Ok looking at this method I'd recommend a few things for code clarity:

  * Drop the last parameter here
  * Drop the `if/else` method in here
  * Where we call this method, do:
 
```java
if (singleProvider) {
   PrivacyUtils.blah()
} else {
    termsText.setText(multipleMessage)
}
```

I think that's easier to read and more similar to how we use the non-SMS methods.

File: auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+                    R.string.fui_verify_phone_number,
+                    getTermsSmsStringResource(flowParameters),
+                    termsText);
+        } else {
+            termsText.setText(multipleProviderFlowSmsMsg);
+        }
+    }
+
+    private static @StringRes
+    int getGlobalTermsStringResource(FlowParameters flowParameters) {
```
Comment (samtstern) [inline auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java]: Formatting nit (here and elsewhere) keep the whole method declaration up to the arg list on one line:

```java
private static @StringRes int getGlobalTermsStringResource(FlowParameters flowParameters) {
```

If you need to line break, do it after the `(`

File: auth/src/main/java/com/firebase/ui/auth/util/ui/PreambleHandler.java (hunk 1 above)
Comment (samtstern) [reply auth/src/main/java/com/firebase/ui/auth/util/ui/PreambleHandler.java]: Bump.

File: auth/src/main/res/layout-land/fui_auth_method_picker_layout.xml [hunk 23]
```diff
@@ -34,4 +34,13 @@
 
     </ScrollView>
 
+    <TextView
+        android:id="@+id/main_tos_and_pp"
+        style="@style/FirebaseUI.Text.BodyText"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="@dimen/fui_field_padding_vert"
+        android:textColor="?android:textColorTertiary"
```
Comment (samtstern) [inline auth/src/main/res/layout-land/fui_auth_method_picker_layout.xml]: What if you do `?attr:textColorTertiary` does it look the same?  If so, prefer that form since it will be more theme-able.

File: auth/src/main/res/layout/fui_auth_method_picker_layout.xml (hunk 10 above)
Comment (samtstern) [reply auth/src/main/res/layout/fui_auth_method_picker_layout.xml]: My comment above addresses this I think.

File: auth/src/main/res/layout/fui_auth_method_picker_layout.xml [hunk 24]
```diff
@@ -26,12 +27,23 @@
...
+        android:textIsSelectable="false"
+        android:gravity="center"
+        app:layout_constraintTop_toBottomOf="@id/container"/>
```
Comment (samtstern) [inline auth/src/main/res/layout/fui_auth_method_picker_layout.xml]: Have you tested this with all numbers of buttons?  IMO it would be more natural for this to be constrained `Bottom_toBottomOf="parent"` so it's always exactly on the bottom.

Also make sure it looks OK in landscape.

File: auth/src/main/res/layout/fui_phone_layout.xml [hunk 25]
```diff
@@ -1,53 +1,70 @@
...
+                android:layout_marginTop="@dimen/fui_field_padding_vert"
+                android:textColor="?android:textColorTertiary"
+                android:textIsSelectable="false"
+                app:layout_constraintTop_toBottomOf="@+id/send_code" />
+
+        </android.support.constraint.ConstraintLayout>
+
+    </ScrollView>
+
+    <TextView
```
Comment (samtstern) [inline auth/src/main/res/layout/fui_phone_layout.xml]: We repeat this a lot, we could extract most of the attributes into a style and then this would only be like a two-line copy paste (ID and style) into each layout instead of a ~10 line.

File: auth/src/main/res/layout/fui_phone_layout.xml [hunk 26]
```diff
@@ -1,53 +1,70 @@
...
+        android:id="@+id/email_footer_tos_and_pp_text"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="@dimen/fui_field_padding_vert"
+        android:paddingRight="15dp"
```
Comment (samtstern) [inline auth/src/main/res/layout/fui_phone_layout.xml]: nit: 15 is a strange number, try 16 which is a standard Android padding amount (normally you work in multiples of 4dp)

File: auth/src/main/res/values/strings.xml (hunk 11 above)
Comment (samtstern) [reply auth/src/main/res/values/strings.xml]: Sent off the strings for translation.

File: auth/src/main/java/com/firebase/ui/auth/ui/email/CheckEmailFragment.java (hunk 17 above)
Comment (SUPERCILEX) [reply auth/src/main/java/com/firebase/ui/auth/ui/email/CheckEmailFragment.java]: Oh yeah, I forgot to mention something else about this. You don't need those casts! Notice how the IDE greys them out? That's cuz `findViewById` was updated to use generics in API 26.

File: auth/src/main/java/com/firebase/ui/auth/ui/phone/SubmitConfirmationCodeFragment.java (hunk 19 above)
Comment (lsirac) [reply auth/src/main/java/com/firebase/ui/auth/ui/phone/SubmitConfirmationCodeFragment.java]: yep

File: auth/src/main/java/com/firebase/ui/auth/ui/idp/AuthMethodPickerActivity.java (hunk 18 above)
Comment (lsirac) [reply auth/src/main/java/com/firebase/ui/auth/ui/idp/AuthMethodPickerActivity.java]: I'll change to this

File: auth/src/main/java/com/firebase/ui/auth/ui/phone/VerifyPhoneNumberFragment.java (hunk 20 above)
Comment (lsirac) [reply auth/src/main/java/com/firebase/ui/auth/ui/phone/VerifyPhoneNumberFragment.java]: Yes, SMS is special because we warn them either way (about the SMS cost).

File: auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java (hunk 21 above)
Comment (lsirac) [reply auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java]: I agree. Changed it.

File: auth/src/main/res/layout/fui_auth_method_picker_layout.xml (hunk 10 above)
Comment (lsirac) [reply auth/src/main/res/layout/fui_auth_method_picker_layout.xml]: Changed it to attr/textColorTertiary

File: auth/src/main/res/layout/fui_phone_layout.xml (hunk 25 above)
Comment (lsirac) [reply auth/src/main/res/layout/fui_phone_layout.xml]: We'll do this in a different PR if that's OK

**Comment by samtstern** [comment]
@lsirac we've got some new lint errors

File: library/quality/quality.gradle (hunk 16 above)
Comment (lsirac) [reply library/quality/quality.gradle]: Done, thanks!

File: auth/src/main/java/com/firebase/ui/auth/ui/email/CheckEmailFragment.java [hunk 27]
```diff
@@ -96,6 +99,21 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
         }
 
         view.findViewById(R.id.button_next).setOnClickListener(this);
+
+        TextView termsText = view.findViewById(R.id.email_tos_and_pp_text);
+        TextView footerText = view.findViewById(R.id.email_footer_tos_and_pp_text);
+        FlowParameters flowParameters = getFlowParams();
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/ui/email/CheckEmailFragment.java]: Can we get rid of `flowParameters`? (It's just as long as `getFlowParams()` and is only used once.)

File: auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java [hunk 28]
```diff
@@ -0,0 +1,86 @@
...
+import com.firebase.ui.auth.R;
+import com.firebase.ui.auth.data.model.FlowParameters;
+import com.firebase.ui.auth.util.ui.PreambleHandler;
+
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+public class PrivacyDisclosureUtils {
+
+    private static final int NO_TOS_OR_PP = -1;
+
+    public static void setupTermsOfServiceAndPrivacyPolicyText(Context context,
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java]: Wondering if we should shorten this to `setupTosAndPp`. Or at least, `setupTermsOfServiceAndPrivacyPolicyText` is quite a mouthful. 😆 @samtstern Thoughts?

File: auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java [hunk 29]
```diff
@@ -0,0 +1,86 @@
...
+        boolean termsOfServiceUrlProvided = flowParameters.isTermsOfServiceUrlProvided();
+        boolean privacyPolicyUrlProvided = flowParameters.isPrivacyPolicyUrlProvided();
+        if (termsOfServiceUrlProvided && privacyPolicyUrlProvided) {
+            return R.string.fui_tos_and_pp_footer;
+        } else if (termsOfServiceUrlProvided) {
+            return R.string.fui_tos_footer;
+        } else if (privacyPolicyUrlProvided) {
+            return R.string.fui_pp_footer;
+        }
+        return NO_TOS_OR_PP; // add constant
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java]: > // add constant

Looks like you did that, comment no longer needed.

File: auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java [hunk 30]
```diff
@@ -0,0 +1,86 @@
...
+                R.string.fui_verify_phone_number,
+                getTermsSmsStringResource(flowParameters),
+                termsText);
+    }
+
+    @StringRes
+    private static int getGlobalTermsStringResource(FlowParameters flowParameters) {
+        boolean termsOfServiceUrlProvided = flowParameters.isTermsOfServiceUrlProvided();
+        boolean privacyPolicyUrlProvided = flowParameters.isPrivacyPolicyUrlProvided();
+        if (termsOfServiceUrlProvided && privacyPolicyUrlProvided) {
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java]: Can we add a new line before and after the if/else here and in the other methods. Makes it more readable IMO.

File: auth/src/main/res/layout/fui_check_email_layout.xml [hunk 31]
```diff
@@ -1,38 +1,57 @@
...
+            <TextView
+                android:id="@+id/email_tos_and_pp_text"
+                style="@style/FirebaseUI.Text.BodyText"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:layout_marginTop="@dimen/fui_field_padding_vert"
+                android:textIsSelectable="false"
+                tools:text="Terms and Privacy Policy" />
+
+            <TextView
```
Comment (SUPERCILEX) [inline auth/src/main/res/layout/fui_check_email_layout.xml]: What's the difference between this TextView and the one above? If this isn't a mistake, we should add a comment explaining that.

**Comment by SUPERCILEX** [comment]
Oh hey, looks like @samtstern fixed them!

**Comment by samtstern** [comment]
@SUPERCILEX the layouts **are** pointless but they're gonna need to be there for my progress dialog thing anyway so just making my merge life easier.

File: auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java (hunk 30 above)
Comment (lsirac) [reply auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java]: Done.

File: auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java (hunk 29 above)
Comment (lsirac) [reply auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java]: :no_mouth: done.

File: auth/src/main/res/layout/fui_check_email_layout.xml (hunk 31 above)
Comment (lsirac) [reply auth/src/main/res/layout/fui_check_email_layout.xml]: Done.

File: auth/src/main/java/com/firebase/ui/auth/ui/email/CheckEmailFragment.java (hunk 27 above)
Comment (lsirac) [reply auth/src/main/java/com/firebase/ui/auth/ui/email/CheckEmailFragment.java]: Done.

File: auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java (hunk 28 above)
Comment (lsirac) [reply auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java]: I prefer to keep it like this :blush:, do you both want me to change it? @samtstern

File: auth/src/main/java/com/firebase/ui/auth/data/remote/SignInKickstarter.java [hunk 32]
```diff
@@ -92,9 +92,8 @@ public void onComplete(@NonNull Task<CredentialRequestResponse> task) {
 
     private void startAuthMethodChoice() {
         List<AuthUI.IdpConfig> idpConfigs = getArguments().providerInfo;
```
[the remark this answers is not in the document]
Comment (lsirac) [reply auth/src/main/java/com/firebase/ui/auth/data/remote/SignInKickstarter.java]: Not sure what you mean. Ctrl + Shift + N lets me search for files. 

I changed it so that there is no local idpConfigs list, we just do getArguments().providerInfo.get(0) instead. Should we add a sanity check here since we access the first element without checking if it exists?

File: auth/src/main/java/com/firebase/ui/auth/ui/email/CheckEmailFragment.java (hunk 27 above)
Comment (SUPERCILEX) [reply auth/src/main/java/com/firebase/ui/auth/ui/email/CheckEmailFragment.java]: Nope, SGTM

File: auth/src/main/java/com/firebase/ui/auth/data/remote/SignInKickstarter.java (hunk 32 above)
[the remark this answers is not in the document]
Comment (SUPERCILEX) [reply auth/src/main/java/com/firebase/ui/auth/data/remote/SignInKickstarter.java]: Lol, my B. I meant `Ctrl + Alt + N`. 😊 You should try it, I'm always impressed.

[the remark this answers is not in the document]
Comment (SUPERCILEX) [reply auth/src/main/java/com/firebase/ui/auth/data/remote/SignInKickstarter.java]: But yeah, LGTM

File: auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java (hunk 28 above)
Comment (SUPERCILEX) [reply auth/src/main/java/com/firebase/ui/auth/util/data/PrivacyDisclosureUtils.java]: If you think that's clearer then SGTM 👍.

**Comment by SUPERCILEX** [comment]
Alrighty, waiting for you to push!

**Comment by samtstern** [comment]
Merged, welcome to the contributors list @lsirac!
